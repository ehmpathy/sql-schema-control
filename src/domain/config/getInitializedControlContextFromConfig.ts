/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the language+dialect, connection, and definitions
*/
