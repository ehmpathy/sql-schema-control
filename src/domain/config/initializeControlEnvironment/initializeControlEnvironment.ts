/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
