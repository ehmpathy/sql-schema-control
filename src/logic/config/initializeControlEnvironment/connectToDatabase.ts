import mysql from 'mysql2/promise';
import { DatabaseConnection, DatabaseLanguage, ControlConfig, ConnectionConfig } from '../../../types';

// create the connection to database
const connectionAdapters = {
  [DatabaseLanguage.MYSQL]: async ({ connectionConfig }: { connectionConfig: ConnectionConfig }): Promise<DatabaseConnection> => {
    const connection = await mysql.createConnection({
      host: connectionConfig.host,
      port: connectionConfig.port,
      database: connectionConfig.schema,
      user: connectionConfig.username,
      password: connectionConfig.password,
    });
    return {
      query: async ({ sql, values }: { sql: string, values?: any[] }) => {
        return connection.query(sql, values);
      },
      end: async () => connection.end(),
    };
  },
};

export const connectToDatabase = async ({ config }: { config: ControlConfig }): Promise<DatabaseConnection> => {
  // 1. get the connection adapter for the method
  const getDbConnection = connectionAdapters[config.language];

  // 2. attempt to connect
  const connection = await getDbConnection({ connectionConfig: config.connection });

  // 3. run a test query to check connection
  await connection.query({ sql: 'SHOW TABLES' }); // TODO: make this not mysql specific

  // 4. return the connection
  return connection;
};
