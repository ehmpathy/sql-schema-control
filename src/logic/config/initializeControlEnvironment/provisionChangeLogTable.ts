import { DatabaseConnection } from '../../../types';
import { readFileAsync } from '../_utils/readFileAsync';

export const provisionChangeLogTable = async ({ connection }: { connection: DatabaseConnection }) => {
  // create the table if not exists (sql consideres its existance for us)
  const createTableSql = await readFileAsync({
    filePath: `${__dirname}/../../../../schema/tables/schema_control_change_log.mysql.sql`,
  }); // TODO: make independent of language (mysql currently)
  await connection.query({ sql: createTableSql });
};
