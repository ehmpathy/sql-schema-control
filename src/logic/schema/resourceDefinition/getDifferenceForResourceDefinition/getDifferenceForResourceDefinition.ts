import { getLiveResourceDefinitionFromDatabase } from '../getLiveResourceDefinitionFromDatabase';
import { DatabaseConnection, ResourceDefinition } from '../../../../types';
import { getSqlDifference } from '../../utils/getSqlDifference';
import { stripIrrelevantContentFromResourceDDL } from './stripIrrelevantContentFromResourceDDL/stripIrrelevantContentFromResourceDDL';

export const getDifferenceForResourceDefinition = async ({
  connection,
  resource,
}: {
  connection: DatabaseConnection;
  resource: ResourceDefinition;
}): Promise<string | null> => {
  // 1. get state of change in db
  const liveResource = await getLiveResourceDefinitionFromDatabase({
    connection,
    resourceName: resource.name,
    resourceType: resource.type,
  });

  // 2. cast into string
  const oldSql = stripIrrelevantContentFromResourceDDL({ ddl: liveResource.sql, resourceType: liveResource.type });
  const newSql = stripIrrelevantContentFromResourceDDL({ ddl: resource.sql, resourceType: resource.type });
  const sqlDifference = getSqlDifference({
    // trim and add newline to standardize how diff calculates and displays
    oldSql: `${oldSql.trim()}\n`,
    newSql: `${newSql.trim()}\n`,
  });

  // 3. return the formatted difference
  return sqlDifference;
};
