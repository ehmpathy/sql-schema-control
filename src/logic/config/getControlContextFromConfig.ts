import { ControlContext, ResourceDefinition } from '../../types';
import { getUncontrolledResources } from '../schema/resourceDefinition/getUncontrolledResources';
import { getConfig } from './getConfig';
import { initializeControlEnvironment } from './initializeControlEnvironment';

/*
  1. get the config (validated and hydrated)
  2. initialize the environment based on the config
    - start the db connection
    - initialize the control environment
  3. return the control context
*/
export const getControlContextFromConfig = async ({
  configPath,
  strict = false,
}: {
  configPath: string;
  strict?: boolean;
}) => {
  // 1. get the config
  const config = await getConfig({ configPath });

  // 2. initialize the environment based on the config
  const { connection } = await initializeControlEnvironment({ config });

  // 3. define the definitions such that we can append to them if we want (e.g., if want to add uncontrolled)
  const definitions = [...config.definitions];

  // 4. get uncontrolled resources, if strict
  if (config.strict || strict) {
    const controlledResources = definitions.filter(
      (def) => def.constructor === ResourceDefinition,
    ) as ResourceDefinition[];
    const uncontrolledResources = await getUncontrolledResources({ connection, controlledResources });
    definitions.push(...uncontrolledResources); // append the uncontrolled resources to the definitions
  }

  // 5. return the control context
  return new ControlContext({
    language: config.language,
    dialect: config.dialect,
    connection, // note: the connection is a live dbConnection
    definitions, // note: uncontrolled resources will already have statuses
  });
};
