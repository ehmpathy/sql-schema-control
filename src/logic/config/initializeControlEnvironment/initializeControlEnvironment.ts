import { ControlConfig } from '../../../domain';
import { connectToDatabase } from './connectToDatabase';
import { provisionChangeLogTable } from './provisionChangeLogTable';

/*
  complete any actions required to initialize control environment
    1. connect to db
    2. provision change log table
*/
export const initializeControlEnvironment = async ({
  config,
}: {
  config: ControlConfig;
}) => {
  // 1. connect to db
  const connection = await connectToDatabase({ config });

  // 2. provision the change log table
  await provisionChangeLogTable({ connection });

  // 3. return the connection
  return { connection };
};
