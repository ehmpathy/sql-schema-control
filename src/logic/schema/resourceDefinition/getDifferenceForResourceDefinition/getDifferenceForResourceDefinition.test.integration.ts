import {
  ControlConfig,
  DatabaseConnection,
  DatabaseLanguage,
  ResourceDefinition,
  ResourceDefinitionStatus,
  ResourceType,
} from '../../../../types';
import { initializeControlEnvironment } from '../../../config/initializeControlEnvironment';
import { getDifferenceForResourceDefinition } from './getDifferenceForResourceDefinition';
import { promiseConfig } from '../../../../__test_assets__/connection.config';

/**
 * note: these are arguably the most important tests - because they define how useful "resources" really are
 *
 * the main advantage of resources is being able to easily see diff between what's checked into version control and whats deployed
 *
 * this is proof that it works
 */
describe('getDifferenceForResourceDefinition', () => {
  describe('mysql', () => {
    let connection: DatabaseConnection;
    beforeAll(async () => {
      const config = new ControlConfig({
        language: DatabaseLanguage.MYSQL,
        dialect: '5.7',
        connection: (await promiseConfig()).mysql,
        definitions: [],
        strict: true,
      });
      ({ connection } = await initializeControlEnvironment({ config })); // ensure db is provisioned and get connection
    });
    afterAll(async () => {
      await connection.end();
    });
    describe('table', () => {
      it('should find that an up to date table has no difference', async () => {
        await connection.query({ sql: 'DROP TABLE IF EXISTS some_resource_table' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.TABLE,
          name: 'some_resource_table',
          path: '__PATH__',
          sql: `
CREATE TABLE \`some_resource_table\` (
  \`id\` int(11) NOT NULL AUTO_INCREMENT,
  \`created_at\` datetime DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (\`id\`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
          `.trim(),
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: definition });
        expect(result).toEqual(null);
      });
      it('should find that an up to date table has no difference, even if auto increment has gone up', async () => {
        await connection.query({ sql: 'DROP TABLE IF EXISTS some_resource_table' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.TABLE,
          name: 'some_resource_table',
          path: '__PATH__',
          sql: `
CREATE TABLE \`some_resource_table\` (
  \`id\` int(11) NOT NULL AUTO_INCREMENT,
  \`created_at\` datetime DEFAULT CURRENT_TIMESTAMP,
  PRIMARY KEY (\`id\`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
          `.trim(),
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // insert into the table
        await connection.query({ sql: 'INSERT INTO some_resource_table (created_at) VALUES (now(6))' });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: definition });
        expect(result).toEqual(null);
      });
      it('should find that an out of date table resource has a colored diff of changed lines', async () => {
        await connection.query({ sql: 'DROP TABLE IF EXISTS some_resource_table' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.TABLE,
          name: 'some_resource_table',
          path: '__PATH__',
          sql: `
        CREATE TABLE \`some_resource_table\` (
          \`id\` INT(11) AUTO_INCREMENT,
          \`created_at\` DATETIME DEFAULT CURRENT_TIMESTAMP,
          PRIMARY KEY (\`id\`)
          ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
          `,
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // update the original definition
        const updatedDefinition = new ResourceDefinition({
          ...definition,
          sql: `
          CREATE TABLE \`some_resource_table\` (
            \`id\` BIGINT AUTO_INCREMENT,
            \`created_at\` datetime DEFAULT CURRENT_TIMESTAMP,
            PRIMARY KEY (\`id\`)
            ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
            `,
        });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: updatedDefinition });
        expect(typeof result).toEqual('string');
        expect(result).toMatchSnapshot(); // result is purely visual, so log an example of it
      });
    });
    describe('procedure', () => {
      it('should find that an up to date procedure has no difference', async () => {
        await connection.query({ sql: 'DROP PROCEDURE IF EXISTS upsert_user_description' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.PROCEDURE,
          name: 'upsert_user_description',
          path: '__PATH__',
          sql: `
CREATE PROCEDURE \`upsert_user_description\`(
  IN in_from_user_id BIGINT
)
BEGIN
  -- just select something, and also include a comment
  SELECT false;
END
      `,
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: definition });
        expect(result).toEqual(null);
      });
      it('should find that an out of date procedure resource has a colored diff of changed lines', async () => {
        await connection.query({ sql: 'DROP PROCEDURE IF EXISTS upsert_user_description' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.PROCEDURE,
          name: 'upsert_user_description',
          path: '__PATH__',
          sql: `
CREATE PROCEDURE upsert_user_description(
  IN in_from_user_id BIGINT
)
BEGIN
  -- just select something, and also include a comment
  SELECT false;
END;
      `,
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // update the original definition
        const updatedDefinition = new ResourceDefinition({
          ...definition,
          sql: `
CREATE PROCEDURE upsert_user_description(
  IN in_from_user_id BIGINT
)
BEGIN
  -- just select something, and also include a comment
  SELECT true;
END
      `,
        });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: updatedDefinition });
        expect(typeof result).toEqual('string');
        expect(result).toMatchSnapshot(); // result is purely visual, so log an example of it
      });
    });
    describe('view', () => {
      it('should find that an up to date view has no difference', async () => {
        await connection.query({ sql: 'DROP VIEW IF EXISTS view_spacecraft_current' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.VIEW,
          name: 'view_spacecraft_current',
          path: '__PATH__',
          sql: `
CREATE VIEW view_spacecraft_current AS
SELECT
  'starship 7' as name; -- NOTE: hard coded so we don't have to provision tables; we test this better in the 'normalizeDDLToSupportLossyShowCreateStatements'
      `,
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: definition });
        expect(result).toEqual(null);
      });
      it('should find that an out of date view resource has a colored diff of changed lines', async () => {
        await connection.query({ sql: 'DROP VIEW IF EXISTS view_spacecraft_current' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.VIEW,
          name: 'view_spacecraft_current',
          path: '__PATH__',
          sql: `
CREATE VIEW view_spacecraft_current AS
SELECT
  'starship 7' as name;
      `,
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // update the original definition
        const updatedDefinition = new ResourceDefinition({
          ...definition,
          sql: `
CREATE VIEW view_spacecraft_current AS
SELECT
  'starship 7' as ship_name;
      `,
        });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: updatedDefinition });
        expect(typeof result).toEqual('string');
        expect(result).toMatchSnapshot(); // result is purely visual, so log an example of it
      });
    });
  });
  describe('postgres', () => {
    let connection: DatabaseConnection;
    beforeAll(async () => {
      const config = new ControlConfig({
        language: DatabaseLanguage.POSTGRES,
        dialect: '5.7',
        connection: (await promiseConfig()).postgres,
        definitions: [],
        strict: true,
      });
      ({ connection } = await initializeControlEnvironment({ config })); // ensure db is provisioned and get connection
    });
    afterAll(async () => {
      await connection.end();
    });
    describe('table', () => {
      it('should find that an up to date table has no difference', async () => {
        await connection.query({ sql: 'DROP TABLE IF EXISTS some_resource_table' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.TABLE,
          name: 'some_resource_table',
          path: '__PATH__',
          sql: `
CREATE TABLE some_resource_table (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  CONSTRAINT some_resource_table_pk PRIMARY KEY (id)
);
          `.trim(),
          status: undefined,
        });
        await connection.query({ sql: definition.sql });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: definition });
        expect(result).toEqual(null);
      });
      it('should find that an up to date table has no difference, even if auto increment has gone up', async () => {
        await connection.query({ sql: 'DROP TABLE IF EXISTS some_resource_table' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.TABLE,
          name: 'some_resource_table',
          path: '__PATH__',
          sql: `
CREATE TABLE some_resource_table (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  CONSTRAINT some_resource_table_pk PRIMARY KEY (id)
);
          `.trim(),
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // insert into the table
        await connection.query({ sql: 'INSERT INTO some_resource_table (created_at) VALUES (now())' });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: definition });
        expect(result).toEqual(null);
      });
      it('should find that an out of date table resource has a colored diff of changed lines', async () => {
        await connection.query({ sql: 'DROP TABLE IF EXISTS some_resource_table' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.TABLE,
          name: 'some_resource_table',
          path: '__PATH__',
          sql: `
CREATE TABLE some_resource_table (
  id bigserial NOT NULL,
  created_at timestamp with time zone NOT NULL DEFAULT now(),
  CONSTRAINT some_resource_table_pk PRIMARY KEY (id)
)
          `.trim(),
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // update the original definition
        const updatedDefinition = new ResourceDefinition({
          ...definition,
          sql: `
CREATE TABLE some_resource_table(
  id serial,
  created_at timestamp DEFAULT now(),
  CONSTRAINT some_resource_table_pk PRIMARY KEY (id)
)
          `.trim(),
        });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: updatedDefinition });
        expect(typeof result).toEqual('string');
        expect(result).toMatchSnapshot(); // result is purely visual, so log an example of it
      });
    });
    describe('function', () => {
      it('should find that an up to date function has no difference', async () => {
        await connection.query({ sql: 'DROP FUNCTION IF EXISTS upsert_user_description' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.FUNCTION,
          name: 'upsert_user_description',
          path: '__PATH__',
          sql: `
CREATE OR REPLACE FUNCTION upsert_user_description(
  in_from_user_id bigint
)
RETURNS boolean
LANGUAGE plpgsql
AS $$
  BEGIN
    -- just select something, and also include a comment
    SELECT false;
  END;
$$
      `,
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: definition });
        expect(result).toEqual(null);
      });
      it('should find that an out of date function resource has a colored diff of changed lines', async () => {
        await connection.query({ sql: 'DROP FUNCTION IF EXISTS upsert_user_description' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.FUNCTION,
          name: 'upsert_user_description',
          path: '__PATH__',
          sql: `
CREATE OR REPLACE FUNCTION upsert_user_description(
  in_from_user_id bigint
)
RETURNS boolean
LANGUAGE plpgsql
AS $$
  BEGIN
    -- just select something, and also include a comment
    SELECT false;
  END;
$$
      `,
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // update the original definition
        const updatedDefinition = new ResourceDefinition({
          ...definition,
          sql: `
CREATE OR REPLACE FUNCTION upsert_user_description(
  in_from_user_id bigint,
  in_another_arg boolean
)
RETURNS boolean
LANGUAGE plpgsql
AS $$
  BEGIN
    -- just select something, and also include a comment, which we changed here
    SELECT false;
  END;
$$
      `,
        });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: updatedDefinition });
        expect(typeof result).toEqual('string');
        expect(result).toMatchSnapshot(); // result is purely visual, so log an example of it
      });
    });
    describe('view', () => {
      it('should find that an up to date view has no difference', async () => {
        await connection.query({ sql: 'DROP VIEW IF EXISTS view_spacecraft_current' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.VIEW,
          name: 'view_spacecraft_current',
          path: '__PATH__',
          sql: `
CREATE OR REPLACE VIEW view_spacecraft_current AS
SELECT
  'starship 7' as name; -- NOTE: hard coded so we don't have to provision tables; we test this better in the 'normalizeDDLToSupportLossyShowCreateStatements'
      `,
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: definition });
        expect(result).toEqual(null);
      });
      it('should find that an out of date view resource has a colored diff of changed lines', async () => {
        await connection.query({ sql: 'DROP VIEW IF EXISTS view_spacecraft_current' });

        // apply the resource
        const definition = new ResourceDefinition({
          type: ResourceType.VIEW,
          name: 'view_spacecraft_current',
          path: '__PATH__',
          sql: `
CREATE OR REPLACE VIEW view_spacecraft_current AS
SELECT
  'starship 7' as name;
      `,
          status: ResourceDefinitionStatus.OUT_OF_SYNC,
        });
        await connection.query({ sql: definition.sql });

        // update the original definition
        const updatedDefinition = new ResourceDefinition({
          ...definition,
          sql: `
CREATE OR REPLACE VIEW view_spacecraft_current AS
SELECT
  'starship 7' as ship_name;
      `,
        });

        // get the diff
        const result = await getDifferenceForResourceDefinition({ connection, resource: updatedDefinition });
        expect(typeof result).toEqual('string');
        expect(result).toMatchSnapshot(); // result is purely visual, so log an example of it
      });
    });
  });
});
