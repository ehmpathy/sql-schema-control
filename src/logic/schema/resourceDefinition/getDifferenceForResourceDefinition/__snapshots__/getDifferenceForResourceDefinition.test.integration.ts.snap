// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDifferenceForResourceDefinition procedure should find that an out of date procedure resource has a colored diff of changed lines 1`] = `
"[31mCREATE PROCEDURE \`upsert_user_description\`([39m
[31m[39m[32mCREATE PROCEDURE upsert_user_description([39m
[32m[39m[90mIN in_from_user_id BIGINT[39m
[90m)[39m
[90mBEGIN[39m
[90m-- just select something, and also include a comment[39m
[90m[39m[31mSELECT false;[39m
[31m[39m[32mSELECT true;[39m
[32m[39m[90mEND[39m
[90m[39m"
`;

exports[`getDifferenceForResourceDefinition table should find that an out of date table resource has a colored diff of changed lines 1`] = `
"[90mCREATE TABLE \`some_resource_table\` ([39m
[90m[39m[31m\`id\` int(11) NOT NULL AUTO_INCREMENT,[39m
[31m[39m[32m\`id\` BIGINT AUTO_INCREMENT,[39m
[32m[39m[90m\`created_at\` datetime DEFAULT CURRENT_TIMESTAMP,[39m
[90mPRIMARY KEY (\`id\`)[39m
[90m) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci[39m
[90m[39m"
`;
